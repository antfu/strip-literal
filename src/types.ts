export interface StripLiteralOptions {
  /**
   * Will be called for each string literal. Return false to skip stripping.
   */
  filter?: (s: string) => boolean
}
