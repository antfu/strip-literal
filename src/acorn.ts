import { tokenizer } from 'acorn'

/**
 * Strip literal using Acorn's tokenizer.
 *
 * Will throw error if the input is not valid JavaScript.
 */
export function stripLiteralAcorn(code: string) {
  const FILL = ' '
  let result = ''
  function fulfill(index: number) {
    if (index > result.length)
      result += code.slice(result.length, index).replace(/[^\n]/g, FILL)
  }

  const tokens = tokenizer(code, {
    ecmaVersion: 'latest',
    sourceType: 'module',
    allowHashBang: true,
    allowAwaitOutsideFunction: true,
    allowImportExportEverywhere: true,
  })
  const inter = tokens[Symbol.iterator]()

  while (true) {
    const { done, value: token } = inter.next()
    if (done)
      break
    fulfill(token.start)
    if (token.type.label === 'string')
      result += code[token.start] + FILL.repeat(token.end - token.start - 2) + code[token.end - 1]
    else if (token.type.label === 'template')
      result += FILL.repeat(token.end - token.start)

    else
      result += code.slice(token.start, token.end)
  }

  fulfill(code.length)

  return result
}

/**
 * Returns a function that returns whether the position is
 * in a literal using Acorn's tokenizer.
 *
 * Will throw error if the input is not valid JavaScript.
 */
export function createIsLiteralPositionAcorn(code: string) {
  // literal start position, non-literal start position, literal start position, ...
  const positionList: number[] = []

  const tokens = tokenizer(code, {
    ecmaVersion: 'latest',
    sourceType: 'module',
    allowHashBang: true,
    allowAwaitOutsideFunction: true,
    allowImportExportEverywhere: true,
    onComment(_isBlock, _text, start, end) {
      positionList.push(start)
      positionList.push(end)
    },
  })
  const inter = tokens[Symbol.iterator]()

  while (true) {
    const { done, value: token } = inter.next()
    if (done)
      break
    if (token.type.label === 'string') {
      positionList.push(token.start + 1)
      positionList.push(token.end - 1)
    }
    else if (token.type.label === 'template') {
      positionList.push(token.start)
      positionList.push(token.end)
    }
  }

  return (position: number) => {
    const i = binarySearch(positionList, v => position < v)
    return (i - 1) % 2 === 0
  }
}

function binarySearch(array: ArrayLike<number>, pred: (v: number) => boolean) {
  let low = -1
  let high = array.length
  while (1 + low < high) {
    const mid = low + ((high - low) >> 1)
    if (pred(array[mid]))
      high = mid
    else
      low = mid
  }
  return high
}
