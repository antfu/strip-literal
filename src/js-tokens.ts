import jsTokens, { type Token as JSToken } from 'js-tokens'
import type { StripLiteralOptions } from './types'

export function _stripLiteralJsTokens(code: string, options?: StripLiteralOptions) {
  const FILL = options?.fillChar ?? ' '
  const FILL_COMMENT = ' '
  let result = ''

  const filter = options?.filter ?? (() => true)

  const tokens: JSToken[] = []

  let error: any
  try {
    for (const token of jsTokens(code, { jsx: false })) {
      tokens.push(token)

      if (token.type === 'SingleLineComment') {
        result += FILL_COMMENT.repeat(token.value.length)
        continue
      }

      if (token.type === 'MultiLineComment') {
        result += token.value.replace(/[^\n]/g, FILL_COMMENT)
        continue
      }

      if (token.type === 'StringLiteral') {
        const body = token.value.slice(1, -1)
        if (filter(body)) {
          result += token.value[0] + FILL.repeat(body.length) + token.value[token.value.length - 1]
          continue
        }
      }

      if (token.type === 'NoSubstitutionTemplate') {
        const body = token.value.slice(1, -1)
        if (filter(body)) {
          result += `\`${body.replace(/[^\n]/g, FILL_COMMENT)}\``
          continue
        }
      }

      if (token.type === 'RegularExpressionLiteral') {
        const body = token.value
        if (filter(body)) {
          result += body.replace(/\/(.*)\/(\w?)$/g, (_, $1, $2) => `/${FILL.repeat($1.length)}/${$2}`)
          continue
        }
      }

      // `start${
      if (token.type === 'TemplateHead') {
        const body = token.value.slice(1, -2)
        if (filter(body)) {
          result += `\`${body.replace(/[^\n]/g, FILL_COMMENT)}\${`
          continue
        }
      }

      // }end`
      if (token.type === 'TemplateTail') {
        const body = token.value.slice(0, -2)
        if (filter(body)) {
          result += `}${body.replace(/[^\n]/g, FILL_COMMENT)}\``
          continue
        }
      }

      // }middle${
      if (token.type === 'TemplateMiddle') {
        const body = token.value.slice(1, -2)
        if (filter(body)) {
          result += `}${body.replace(/[^\n]/g, FILL_COMMENT)}\${`
          continue
        }
      }

      result += token.value
    }
  }
  catch (e) {
    error = e
  }

  return {
    error,
    result,
    tokens,
  }
}
