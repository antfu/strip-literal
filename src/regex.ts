import type { StripLiteralOptions } from './types'

const multilineCommentsRE = /\/\*([^*\/])*?\*\//gms
const singlelineCommentsRE = /(?:^|\n|\r)\s*\/\/.*(?:\r|\n|$)/gm
const templateLiteralRE = /\$\{(\s*(?:|{.*}|(?!\$\{).|\n|\r)*?\s*)\}/g
const quotesRE = [
  /(["'`])((?:\\\1|(?!\1)|.|\r)*?)\1/gm,
  /([`])((?:\\\1|(?!\1)|.|\n|\r)*?)\1/gm, // multi-line strings (i.e. template literals only)
]

/**
 * Strip literal using RegExp.
 *
 * This will be faster and can work on non-JavaScript input.
 * But will have some caveats on distinguish strings and comments.
 */
export function stripLiteralRegex(code: string, options?: StripLiteralOptions) {
  const filter = options?.filter ?? (() => true)

  code = code
    .replace(multilineCommentsRE, s => filter(s) ? ' '.repeat(s.length) : s)
    .replace(singlelineCommentsRE, s => filter(s) ? ' '.repeat(s.length) : s)

  let expanded = code
  // Recursively replace ${} to support nested constructs (e.g. ${`${x}`})
  for (let i = 0; i < 16; i++) {
    const before = expanded
    expanded = expanded.replace(templateLiteralRE, '` $1`')
    if (expanded === before)
      break
  }

  quotesRE.forEach((re) => {
    expanded = expanded
      .replace(re, (s, quote, body, index) => {
        if (!filter(s.slice(1, -1)))
          return s
        code = code.slice(0, index + 1) + ' '.repeat(s.length - 2) + code.slice(index + s.length - 1)
        return quote + ' '.repeat(s.length - 2) + quote
      })
  })

  return code
}
