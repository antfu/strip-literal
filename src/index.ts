import { _stripLiteralJsTokens } from './js-tokens'
import { stripLiteralRegex } from './regex'
import type { StripLiteralOptions } from './types'

export { stripLiteralAcorn, createIsLiteralPositionAcorn } from './acorn'
export { stripLiteralRegex } from './regex'
export { _stripLiteralJsTokens } from './js-tokens'
export * from './types'

/**
 * Strip literal from code.
 *
 * Using Acorn's tokenizer first, and fallback to Regex if Acorn fails.
 */
export function stripLiteral(code: string, options?: StripLiteralOptions) {
  return stripLiteralDetailed(code, options).result
}

/**
 * Strip literal from code, return more detailed information.
 *
 * Using Acorn's tokenizer first, and fallback to Regex if Acorn fails.
 */
export function stripLiteralDetailed(code: string, options?: StripLiteralOptions): {
  mode: 'js-tokens' | 'regex'
  result: string
  resultJsTokens: {
    tokens: any[]
    error?: any
  }
} {
  const resultJsTokens = _stripLiteralJsTokens(code, options)
  if (!resultJsTokens.error) {
    return {
      mode: 'js-tokens',
      result: resultJsTokens.result,
      resultJsTokens,
    }
  }
  return {
    mode: 'regex',
    result: stripLiteralRegex(resultJsTokens.result + code.slice(resultJsTokens.result.length), options),
    resultJsTokens,
  }
}
