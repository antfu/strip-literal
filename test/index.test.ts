/* eslint-disable no-template-curly-in-string */
import fs from 'fs/promises'
import { parse } from 'acorn'
import { expect, test } from 'vitest'
import { stripLiteral } from '../src'

function executeWithVerify(code: string, verifyAst = true) {
  code = code.trim()
  const result = stripLiteral(code)

  for (let i = 0; i < result.length; i++) {
    if (!result[i].match(/\s/))
      expect(result[i]).toBe(code[i])
  }

  expect(result.length).toBe(code.length)

  // make sure no syntax errors
  if (verifyAst)
    parse(result, { ecmaVersion: 'latest', sourceType: 'module' })

  return result
}

test('works', () => {
  expect(executeWithVerify(`
// comment1
const a = 'aaaa'
/* comment2 */
const b = "bbbb"
/*
  // comment3
*/
/* // comment3 */
// comment 4 /* comment 5 */
const c = \`ccc\${a}\`

let d = /re\\\\ge/g
    `)).toMatchSnapshot()
})

test('escape character', () => {
  expect(executeWithVerify(`
'1\\'1'
"1\\"1"
"1\\"1\\"1"
"1\\'1'\\"1"
"1'1'"
"1'\\'1\\''\\"1\\"\\""
'1"\\"1\\""\\"1\\"\\"'
'""1""'
'"""1"""'
'""""1""""'
"''1''"
"'''1'''"
"''''1''''"
  `)).toMatchSnapshot()
})

test('regexp affect', () => {
  expect(executeWithVerify(`
[
  /'/,
  '1',
  /"/,
  "1"
]
  `)).toMatchSnapshot()
})

test('strings comment nested', () => {
  expect(executeWithVerify(`
// comment 1 /* " */
const a = "a //"
// comment 2 /* " */
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
// comment 1 /* ' */
const a = "a //"
// comment 2 /* ' */
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
// comment 1 /* \` */
const a = "a //"
// comment 2 /* \` */
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
const a = "a //"
console.log("console")
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
const a = "a /*"
console.log("console")
const b = "b */"
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
const a = "a ' "
console.log("console")
const b = "b ' "
  `)).toMatchSnapshot()

  expect(executeWithVerify(`
const a = "a \` "
console.log("console")
const b = "b \` "
  `)).toMatchSnapshot()
})

test('acorn syntax error', () => {
  expect(executeWithVerify(`
foo(\`fooo \${foo({ class: "foo" })} bar\`)
  `, false))
    .toMatchInlineSnapshot('"foo(`     \${foo({ class: \\"   \\" }      `)"')
})

test('template string nested', () => {
  let str = '`aaaa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`    `"')

  str = '`aaaa` `aaaa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`    ` `    `"')

  str = '`aa${a}aa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`  \${a}  `"')

  str = '`aa${a + `a` + a}aa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`  \${a + ` ` + a}  `"')

  str = '`aa${a + `a` + a}aa` `aa${a + `a` + a}aa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`  \${a + ` ` + a}  ` `  \${a + ` ` + a}  `"')

  str = '`aa${a + `aaaa${c + (a = {b: 1}) + d}` + a}aa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`  \${a + `    \${c + (a = {b: 1}) + d}` + a}  `"')

  str = '`aa${a + `aaaa${c + (a = {b: 1}) + d}` + a}aa` `aa${a + `aaaa${c + (a = {b: 1}) + d}` + a}aa`'
  expect(executeWithVerify(str)).toMatchInlineSnapshot('"`  \${a + `    \${c + (a = {b: 1}) + d}` + a}  ` `  \${a + `    \${c + (a = {b: 1}) + d}` + a}  `"')
})

test('backtick escape', () => {
  const str = [
    'this.error(`\\`new URL(url, import.meta.url)\\` is not supported in SSR.`)',
    'this.error(`\\\\`)',
    'this.error(`\\``)',
  ].join('\n')
  expect(executeWithVerify(str)).toMatchInlineSnapshot(`
    "this.error(\`                                                          \`)
    this.error(\`  \`)
    this.error(\`  \`)"
  `)
})

test('forgiving', () => {
  expect(stripLiteral(`
<script type="module">
  const rawModules = import.meta.globEager('/dir/*.json', {
    as: 'raw'
  })
  const globraw = {}
  Object.keys(rawModules).forEach((key) => {
    globraw[key] = JSON.parse(rawModules[key])
  })
  document.querySelector('.globraw').textContent = JSON.stringify(
    globraw,
    null,
    2
  )
</script>
`)).toMatchInlineSnapshot(`
  "
  <script type=\\"      \\">
    const rawModules = import.meta.globEager('           ', {
      as: '   '
    })
    const globraw = {}
    Object.keys(rawModules).forEach((key) => {
      globraw[key] = JSON.parse(rawModules[key])
    })
    document.querySelector('        ').textContent = JSON.stringify(
      globraw,
      null,
      2
    )
  </script>
  "
`)
})

test('// in string', () => {
  const str = [
    'const url= `http://www.xx.com`;',
    'const url1= \'http://www.xx.com\';',
    'onMounted(() => console.log(123))',
    'const str = `hello world`',
    '// Notes',
  ].join('\n')

  expect(executeWithVerify(str)).toMatchInlineSnapshot(`
    "const url= \`                 \`;
    const url1= \'                 \';
    onMounted(() => console.log(123))
    const str = \`           \`
            "
  `)
})

test('#4', async () => {
  const str = (await fs.readFile('./test/fixtures/issue4.ts', 'utf-8'))
    .replace(/\r/g, '') // Windows EOL
  expect(executeWithVerify(str, false)).toMatchInlineSnapshot(`
    "import __variableDynamicImportRuntimeHelper from \\"                          \\";
    const getTestData = async () => {
      const filename = \\"       \\";
      console.log(await __variableDynamicImportRuntimeHelper((import.meta.glob(\\"                       \\")), \`                 \${filename}     \`));
      console.log(await __variableDynamicImportRuntimeHelper((import.meta.glob(\\"                      \\")), \`                \${filename}     \`));
    };

    getTestData();
    const _sfc_main = {};

    import { mergeProps as _mergeProps } from \\"   \\"
    import { ssrRenderAttrs as _ssrRenderAttrs, ssrInterpolate as _ssrInterpolate } from \\"                   \\"

    function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
      _push(\`    \${
        _ssrRenderAttrs(_mergeProps({ class: \\"        \\" }               \${
        _ssrInterpolate(new Date())
      }      \`)
    }


    import { useSSRContext as __vite_useSSRContext } from '   '
    const _sfc_setup = _sfc_main.setup
    _sfc_main.setup = (props, ctx) => {
      const ssrContext = __vite_useSSRContext()
      ;(ssrContext.modules || (ssrContext.modules = new Set())).add(\\"       \\")
      return _sfc_setup ? _sfc_setup(props, ctx) : undefined
    }
    import _export_sfc from '                        '
    export default              _export_sfc(_sfc_main, [['         ',_sfc_ssrRender],['      ',\\"                                                                               \\"]])"
  `)
})
