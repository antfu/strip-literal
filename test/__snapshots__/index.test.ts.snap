// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`escape character 1`] = `
"// mode: js-tokens
'    '
"    "
"       "
"        "
"    "
"               "
'               '
'     '
'       '
'         '
"     "
"       "
"         ""
`;

exports[`regexp affect 1`] = `
"// mode: js-tokens
[
  / /,
  ' ',
  / /,
  " "
]"
`;

exports[`strings comment nested 1`] = `
"// mode: js-tokens
                    
const a = "    "
                    "
`;

exports[`strings comment nested 2`] = `
"// mode: js-tokens
                    
const a = "    "
                    "
`;

exports[`strings comment nested 3`] = `
"// mode: js-tokens
                    
const a = "    "
                    "
`;

exports[`strings comment nested 4`] = `
"// mode: js-tokens
const a = "    "
console.log("       ")"
`;

exports[`strings comment nested 5`] = `
"// mode: js-tokens
const a = "    "
console.log("       ")
const b = "    ""
`;

exports[`strings comment nested 6`] = `
"// mode: js-tokens
const a = "    "
console.log("       ")
const b = "    ""
`;

exports[`strings comment nested 7`] = `
"// mode: js-tokens
const a = "    "
console.log("       ")
const b = "    ""
`;

exports[`works 1`] = `
"// mode: js-tokens
           
const a = '    '
              
const b = "    "
  
             
  
                 
                            
const c = \`   \${a}\`

let d = /      /g"
`;
